programs/token/src/state/mint.rs

```diff
@@ -0,0 +1,91 @@
+use crate::ID;
+
+use pinocchio::{account_info::AccountInfo, program_error::ProgramError, pubkey::Pubkey};
+pub struct Mint(*const u8);
+
+impl Mint {
+    pub const LEN: usize = 82;
+
+    /// # Safety
+    /// Use this when you don't care to keep track of the borrows &
+    /// the AccountInfo will have the correct owner and account space
+    #[inline(always)]
+    pub unsafe fn from_account_info_unchecked_unsafe(account_info: &AccountInfo) -> Self {
+        Self(account_info.borrow_data_unchecked().as_ptr())
+    }
+
+    /// # Safety
+    /// Use this when you don't care to keep track of the borrows
+    pub unsafe fn from_account_info_unsafe(account_info: &AccountInfo) -> Self {
+        assert_eq!(account_info.data_len(), Self::LEN);
+        assert_eq!(account_info.owner(), &ID);
+        Self::from_account_info_unchecked_unsafe(account_info)
+    }
+
+    /// Use this when you know the AccountInfo will have the correct owner and account space
+    #[inline(always)]
+    pub fn from_account_info_unchecked(account_info: &AccountInfo) -> Result<Self, ProgramError> {
+        let data = account_info.try_borrow_data()?;
+        Ok(Self(data.as_ptr()))
```

Comment:
Actually, maybe all we need is to `map` the `Ref<[u8]>` to a `Ref<Mint>`:
```rust
Ref::map(account_info.try_borrow_data()?, |data| Self(data.as_ptr()))
```
The return type of the method then becomes `Result<Ref<Mint>, ProgramError>`.

Reply:
> Actually, maybe all we need is to `map` the `Ref<[u8]>` to a `Ref<Mint>`:
> 
> ```rust
> Ref::map(account_info.try_borrow_data()?, |data| Self(data.as_ptr()))
> ```
> 
> The return type of the method then becomes `Result<Ref<Mint>, ProgramError>`.

I think it's better too, more convenient for the user

Reply:
Not sure if this is still correct, but I needed to add an unsafe cast because the compiler was yelling at me so I went with this for the code solution at the end:
```rust
pub fn from_account_info(account_info: &AccountInfo) -> Result<Ref<Mint>, ProgramError> {
    if account_info.data_len() != Self::LEN { return Err(ProgramError::InvalidAccountData) }
    if account_info.owner() != &ID { return Err(ProgramError::InvalidAccountData) }
    Ok(Ref::map(account_info.try_borrow_data()?, |data| {
        unsafe { &*(data.as_ptr() as *const Mint) }
    }))
}
 ```